### ISSUE TITLE
Distributed deadlock detection may produce false positives

### ISSUE THREAD

**Issue by metdos** [citusdata/citus#1655]
I observed a failing Travis test on the master branch. Rerunning it, it passed the tests. Talking with @onderkalaci, we should not see any deadlock on single commands, so this is something we should take a look.

Note that, so far we observed this on;
```sql
ALTER TABLE table_name ADD CONSTRAINT constraint_name UNIQUE/PRIMARY
```

Example outputs from Travis;

```
--- /home/travis/build/citusdata/citus/src/test/regress/expected/multi_alter_table_add_constraints.out	2017-09-15 14:14:39.562676004 +0000
+++ /home/travis/build/citusdata/citus/src/test/regress/results/multi_alter_table_add_constraints.out	2017-09-15 14:17:55.286796601 +0000
@@ -120,25 +120,23 @@
 -- Insert row which will conflict with the next unique constraint command
 INSERT INTO unique_test_table VALUES(1, 'Mehmet');
 -- Can not create constraint since it conflicts with the existing data
 ALTER TABLE unique_test_table ADD CONSTRAINT unn_id UNIQUE(id);
 ERROR:  could not create unique index "unn_id_1450035"
 DETAIL:  Key (id)=(1) is duplicated.
 CONTEXT:  while executing command on localhost:57637
 -- Can create unique constraint over multiple columns which must include
 -- distribution column
 ALTER TABLE unique_test_table ADD CONSTRAINT unn_id_name UNIQUE(id, name);
+ERROR:  canceling the transaction since it has involved in a distributed deadlock
 -- Error out, since tables can not have two rows with same id and name.
 INSERT INTO unique_test_table VALUES(1, 'Mehmet');
-ERROR:  duplicate key value violates unique constraint "unn_id_name_1450035"
-DETAIL:  Key (id, name)=(1, Mehmet) already exists.
-CONTEXT:  while executing command on localhost:57638
```

Another example from @velioglu 
```
ALTER TABLE products ADD CONSTRAINT p_key PRIMARY KEY(name);
ERROR:  cannot create constraint on "products"
DETAIL:  Distributed relations cannot have UNIQUE, EXCLUDE, or PRIMARY KEY constraints that do not include the partition column (with an equality operator if EXCLUDE).
ALTER TABLE products ADD CONSTRAINT p_key PRIMARY KEY(product_no);
NOTICE:  using one-phase commit for distributed DDL commands
HINT:  You can enable two-phase commit for extra safety with: SET citus.multi_shard_commit_protocol TO '2pc'
+ERROR:  canceling the transaction since it has involved in a distributed deadlock
INSERT INTO products VALUES(1, 'product_1', 1);
-- Should error out, since we are trying to add a new row having a value on p_key column
-- conflicting with the existing row.
INSERT INTO products VALUES(1, 'product_1', 1);
-ERROR:  duplicate key value violates unique constraint "p_key_1450001"
-DETAIL:  Key (product_no)=(1) already exists.
-CONTEXT:  while executing command on localhost:57638
ALTER TABLE products DROP CONSTRAINT p_key;
+ERROR:  constraint "p_key" of relation "products" does not exist
```

**Comment by onderkalaci** [comment]
Just a note that none of the tests run in parallel with any other tests.

**Comment by marcocitus** [comment]
I wonder whether there are cases in which postgres technically creates a deadlock due to implementation details (e.g. take share lock on A, exclusive lock on B, exclusive lock on A), but can normally handle it by reordering lock queues.

**Comment by onderkalaci** [comment]
> I wonder whether there are cases in which postgres technically creates a deadlock due to implementation details (e.g. take share lock on A, exclusive lock on B, exclusive lock on A), but can normally handle it by reordering lock queues.

Just random thoughts: Assuming that was true, in that case, wouldn't we expect that the command wouldn't finish if we disable postgres deadlock detection (i.e., `set deadlock_timout = '100h'`) given that the deadlock detection re-orders lock queues?

To prove that I tried the following in a loop with setting the timeout to `100h` and never ended up with a deadlock:

```
-- t1 is a distributed table
ALTER TABLE t1 ADD CONSTRAINT p_key PRIMARY KEY(id);
ALTER TABLE t1 DROP CONSTRAINT p_key;
```

We probably need to dig more into this to understand whether your guess is true or not.

**Comment by ozgune** [comment]
I'm quoting an anonymized report from a user to this issue. I don't know if this is relevant.

--
I can reproduce distributed deadlocks reliably with my import script.

1. The script runs a `CREATE TABLE IF NOT EXIST`. Then it distributes the table.
2. Another randomly named table is created to hold raw data. Then that table is distributed as well.
3. Then a `COPY` into that table from csv on STDIN.

**Comment by marcocitus** [comment]
I've seen this occur in Travis on `ALTER TABLE customer_copy_hash ADD COLUMN extra2 INT DEFAULT 0;`.

**Comment by marcocitus** [comment]
Just saw the `ALTER TABLE customer_copy_hash ADD COLUMN extra2 INT DEFAULT 0;` again.

Does this happen on both versions of postgres or only on 10?

**Comment by onderkalaci** [comment]
Some notes on this issue:

I had a chance to discuss this issue with a Postgres expert recently :) He suggested us to check 3 things:

* Autovacuum process might be taking locks, consider ignoring those locks
* Postgres may be acquiring `LOCKTAG_RELATION_EXTEND` type of locks, which shouldn't conflict with any other locks but appears in our lock dumps which could cause the issue
* CitusCloud monitors individual shards by regularly querying them. This might be causing an actual deadlock?


I was able to reliably reproduce these false positive deadlocks with `COPY` (the others are pretty hard to reproduce).

As it turned out, the root source of the `COPY` issue is the `LOCKTAG_RELATION_EXTEND` type of locks. See the [PR 1670](https://github.com/citusdata/citus/pull/1670) for the details.

I'll continue to work on the travis issues more.

(Also note that With gdb, it is not very straightforward to see the locks that led to these false positive deadlocks. The reason is that even if you break once a deadlock is found since there are no real deadlocks, the transactions would be committed on the workers after few msecs. Thus, I've added logs and observed that the relation extension locks are acquired in certain cases.)

**Comment by onderkalaci** [comment]
I'd like to share my further comments here:

As you may have noticed, we've fixed the issue related with concurrent `COPY` operations with #1670. The main source of the false positive deadlock is that Postgres may acquire relation extension locks for a very short duration (i.e., release as soon as the relation is extended on the disk, thus there is no danger of distributed deadlocks with other locks/operations). 

The good news is that for the unexpected deadlocks in the travis, I observed a very similar root cause:
So, in summary, we used to consider relation extension locks on:
* Concurrent `COPY`: Two concurrent worker `COPY` transactions may acquire relation extension lock on the shard itself
*  Add unique/primary keys: We send the queries that create the constraints in parallel for each shard. As each transaction runs, they may acquire relation extension locks on `pg_depend` for adding primary/unique key constraints (i.e., Postgres adds one row to `pg_depend` for the new key)
*  Add column: We send the queries that add the column in parallel for each shard. As each transaction runs, they may acquire relation extension locks on `pg_attribute` for the new column (i.e., Postgres adds one row to `pg_attribute` for the new column)

In all of the above cases, the distributed deadlock detection used to end up finding a self-deadlock since some of the worker transactions are waiting for each other.

## Summary
Now that since we ignore relation extension locks for the distributed deadlock detection purposes (#1670), we should not see the false positive deadlocks anymore. I still find it useful to keep this issue open for some time to see if there are any other such edge cases.

**Comment by sumedhpathak** [comment]
@onderkalaci Any further observed incidences after the fix with #1670?

**Comment by onderkalaci** [comment]
No, I think we're good to close the issue.

**Comment by onderkalaci** [comment]
We've some more edges before closing the issue:

1. Speculative locks (`LOCKTAG_SPECULATIVE_TOKEN`):  These locks are held for a very short duration time and cannot contribute to a deadlock. Speculative locks are used by Postgres for internal notification mechanism among transactions.

2. Tuple locks (`LOCKTAG_TUPLE`): This requires some more investigation. But, the initial impression I had is that these locks are held for a short duration per physical tuple and released before the transaction commits/rollbacks.

3. Autovacuum process: Autovacuum process cancels itself if any modification starts on the table in order to avoid blocking your regular Postgres sessions. That's normal and expected. Thus, any locks held by autovacuum process cannot involve in a distributed deadlock since it'll be released if needed.

We should skip dumping locks with the above cases. One cannot simply reproduce deadlocks with the above locks, but, they become apparent when deadlock_timeout = `1ms` and there is a lot of traffic on the workers.

**Comment by metdos** [comment]
@onderkalaci should we keep open this issue?

**Comment by onderkalaci** [comment]
I think we're good to close this for now. If we see any further false positives, we could open a new issue with the details
